Staggered mesh method for PBC exact exchange

stephen-quiton opened the pull request:
Hello! 

In this PR, I'm adding the staggered mesh method for periodic exact exchange ([J. Chem. Theory Comput. 2024, 20, 18, 7958-7968.](https://doi.org/10.1021/acs.jctc.4c00771 "DOI URL")). This is similar in concept to [kmp2_stagger.py](https://github.com/pyscf/pyscf/blob/master/pyscf/pbc/mp/kmp2_stagger.py) but for reducing the finite-size error from the HF exchange term.

Let us know if we need to modify the PR in any way. 

Thank you,
Stephen

MatthewRHermes commented on pyscf/pbc/scf/khf_stagger.py:
```diff
@@ -0,0 +1,420 @@
+        ecell.a = np.einsum('xi,x->xi', mf.cell.lattice_vectors(), nks)
+        ecell.mesh = np.asarray(mf.cell.mesh) * nks
+
+        print('build_probe_cell: nks ',nks)
```
Could you use the logger instead of "print"?

stephen-quiton replied:
These were debugging print statements that should have been deleted. They're fixed now; thanks!

MatthewRHermes commented on pyscf/pbc/scf/khf_stagger.py:
```diff
@@ -0,0 +1,420 @@
+    Args:
+        kmf (khf.KSCF): KSCF object (KRKS or KRHF supported). For Non-SCF and
+            Split-SCF, the SCF should be converged.
+        type (str, optional): "Regular", "Non-SCF", or "Split-SCF". See
```
Not quite: this actually requires an integer, since you don't call `get_stagger_type_id` in this function. You only call it in the class kernel method.

stephen-quiton replied:
Thank you for the catch! [It should be corrected now.](https://github.com/pyscf/pyscf-forge/blob/82882a9f3e495b31c00ee2373b1baa75f1e73cc0/pyscf/pbc/scf/khf_stagger.py#L35)

MatthewRHermes commented on pyscf/pbc/scf/khf_stagger.py:
```diff
@@ -0,0 +1,420 @@
+        else:
+            self.mf.kernel()
+
+    def kernel(self):
+        self.rerun_scf()
+        results = kernel(self.mf,self.stagger_type,df_type=self.df_type,kshift_rel=self.kshift_rel,nks=self.nks)
+        self.ek = results["E_stagger_M"]
+
+        if isinstance(self.mf,rks.KohnShamDFT):
+            xc = True
+            ni = pbcnumint.KNumInt()
+            _, _, hyb = ni.rsh_and_hybrid_coeff(self.xc, spin=self.cell.spin)
+        elif isinstance(self.mf,khf.KRHF):
+            xc = False
+            self.exc = 0.0
+            hyb = 1.0
+        else:
+            logger.error("KHF Stagger: Invalid SCF type", self.mf.__class__)
+            raise ValueError("Invalid SCF type")
+
+        self.compute_energy_components(hcore=True,nuc=True,j=True,k=False,xc=xc)
+        self.e_tot = hyb * self.ek + self.ehcore + self.enuc + self.ej + self.exc
+        self.dump_flags()
+        return self.e_tot
```
Ideally, we preserve the calling arguments and returns of the parent function. What if I have a guess density matrix I want to use? Etc.

stephen-quiton replied:
Great point, [I've made it so that the user can pass dm0 and other args to kernel,](https://github.com/pyscf/pyscf-forge/blob/82882a9f3e495b31c00ee2373b1baa75f1e73cc0/pyscf/pbc/scf/khf_stagger.py#L340) which will be used for the regular version of staggered mesh.

stephen-quiton commented on pyscf/pbc/scf/khf_stagger.py:
```diff
@@ -0,0 +1,420 @@
+    log = logger.Logger(icell.stdout, verbose)
+    ikpts = kmf.kpts
+
+    def build_probe_cell(mf, nks=None):
```
Good question, I think it's because when I was first implementing these functions, they had variables that were only defined in that `kernel` function (now called `exchange_energy_stagger`), but [I've now moved them inside the `KHF_stagger` class.](https://github.com/pyscf/pyscf-forge/blob/82882a9f3e495b31c00ee2373b1baa75f1e73cc0/pyscf/pbc/scf/khf_stagger.py#L292-L369)

stephen-quiton commented on pyscf/pbc/scf/khf_stagger.py:
```diff
@@ -0,0 +1,420 @@
+        # Compute the Staggered mesh energy
+        Nk = np.prod(nks)
+        E_stagger = -1./Nk * np.einsum('kij,kji', dm_shifted, Kmat) * 0.5
+        E_stagger /= 2.
```
[I've refactored it so it's more compact and readable.](https://github.com/pyscf/pyscf-forge/blob/82882a9f3e495b31c00ee2373b1baa75f1e73cc0/pyscf/pbc/scf/khf_stagger.py#L79-L122) Thanks for pointing this out!

stephen-quiton commented on pyscf/pbc/scf/khf_stagger.py:
```diff
@@ -0,0 +1,420 @@
+from pyscf.pbc.tools.pbc import get_monkhorst_pack_size
+
+
+def kernel(kmf, type="Non-SCF", df_type=None, kshift_rel=0.5, verbose=logger.NOTE, with_vk=False, nks=None):
```
Makes sense, [changed to `exchange_energy_stagger`](https://github.com/pyscf/pyscf-forge/blob/82882a9f3e495b31c00ee2373b1baa75f1e73cc0/pyscf/pbc/scf/khf_stagger.py#L27)

stephen-quiton commented:
Thank you for the feedback @MatthewRHermes! I believe I've addressed your comments, but please let me know if any other changes need to be made or if I missed anything.

sunqm commented:
The error in CI is very strange. Can you merge the head?

stephen-quiton commented:
> The error in CI is very strange. Can you merge the head?
Done!
